presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,265 @@
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query materializedViewQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+        this.materializedViewQuery = requireNonNull(materializedViewQuery, "materialized view query is null");
+    }
+
+    public Node rewrite(Node node)
+    {
+        try {
+            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
+            materializedViewInformationExtractor.process(materializedViewQuery);
+            this.materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
+            return process(node);
+        }
+        catch (Exception ex) {
+            logger.error(ex, "Failed to rewrite query: %s.", node);
+            return node;
+        }
+    }
+
+    @Override
+    protected Node visitNode(Node node, Void context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitQuery(Query node, Void context)
+    {
+        QueryBody rewrittenQueryBody = (QueryBody) process(node.getQueryBody(), context);
+        return new Query(
+                node.getWith(),
+                rewrittenQueryBody,
+                node.getOrderBy(),
+                node.getOffset(),
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
+    {
+        if (!node.getFrom().isPresent()) {
+            throw new IllegalStateException("Query with no From clause is not rewritable by materialized view");
+        }
+        Optional<Relation> rewrittenFrom = node.getFrom().map(from -> (Relation) process(from, context));
+        Select rewrittenSelect = (Select) process(node.getSelect(), context);
+
+        // TODO: Handle filter containment problem https://github.com/prestodb/presto/issues/16405
+        if (materializedViewInfo.getWhereClause().isPresent() && !materializedViewInfo.getWhereClause().equals(node.getWhere())) {
+            throw new IllegalStateException("Query with no groupBy clause is not rewritable by materialized view with groupBy clause");
+        }
+        Optional<Expression> rewrittenWhere = node.getWhere().map(where -> (Expression) process(where, context));
+
+        if (materializedViewInfo.getGroupBy().isPresent() && !node.getGroupBy().isPresent()) {
+            throw new IllegalStateException("Query with no groupBy clause is not rewritable by materialized view with groupBy clause");
+        }
+        Optional<GroupBy> rewrittenGroupBy = node.getGroupBy().map(groupBy -> (GroupBy) process(groupBy, context));
+
+        // TODO: Add HAVING validation to the validator https://github.com/prestodb/presto/issues/16406
+        Optional<Expression> rewrittenHaving = node.getHaving().map(having -> (Expression) process(having, context));
```

Comment:
I think we should disable the HAVING queries in the absence of proper validation.